### PR TITLE
Addition of specs and tests for Publisher and Subscriber

### PR DESCRIPTION
**Pull request by rocky02**
This PR adds specs and tests for the Publisher and Subscriber classes of QIt Application.

### CODE REVIEW

File: tests/publisher_spec.rb [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+require_relative '../lib/publisher'
+require_relative '../lib/subscriber'
+
+
+RSpec.describe Subscriber do
+  context '#valid_queue_name?' do
+    let (:valid_queue_name) { "test_123" }
+    
+    it "should validate the url's per aws sqs format" do
+      expect(Publisher.valid_queue_name?(valid_queue_name)).to be_truthy
```
Comment (glen) [inline tests/publisher_spec.rb]: Wasn't there a validation required for the length of the queue name?
Make the changes and test a valid and invalid queue name.

File: lib/subscriber.rb [hunk 2]
```diff
@@ -38,9 +38,6 @@ def self.valid_sleep_period?(sleep_period)
     end
   end
 
-  
-
-# Todo: URL validation /\A(https:\/\/sqs.)#{AWS["region"]}.amazonaws.com\/\d{12}\/([a-zA-Z]*\d*[-_]*)+|(.fifo)\z/
```
Comment (glen) [inline lib/subscriber.rb:43]: We need this validation and test

File: tests/publisher_spec.rb [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    end
+  end
+
+  context '#valid_sleep_period?' do
+    let (:valid_sleep_period) { '7' }
+    it 'subscribes from a queue' do
+      expect(Publisher.valid_sleep_period?(valid_sleep_period)).to be_truthy
```
Comment (glen) [inline tests/publisher_spec.rb]: It's not just success values - also test that invalid sleep period results in failure.

File: tests/publisher_spec.rb [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    end
+  end
+
+  context '#create_std_q' do
+  	let (:queue_name) { 'test123' }
```
Comment (glen) [inline tests/publisher_spec.rb]: Indentation

**Comment by glen** [comment]
Noticed 1 thing - that if there is no `aws.yml` file and I run the specs then just the coverage report runs. Specs should not depend on the `aws.yml` file being present and/or properly filled in with valid values. The idea is that even without user configuration - your specs should work. Understand how you can use a mock for the aws.yml file.

**Comment by glen** [comment]
Also, lets look at having `specs` folder and a `spec_helper.rb` Let stick to convention rather than having someone try to understand how to run the specs out here.

File: lib/subscriber.rb (hunk 2 above)
Comment (rocky02) [reply lib/subscriber.rb:43]: validation has already been added and is in place. So is the test. Needed to add test for invalid URL, which I did now.

File: tests/publisher_spec.rb (hunk 1 above)
Comment (rocky02) [reply tests/publisher_spec.rb]: Length of the queue already part of the regex. {3,80}. So that has been taken care of.

Comment (glen) [reply tests/publisher_spec.rb]: The test for validation of the queue_name is pending.

File: lib/subscriber.rb (hunk 2 above)
Comment (glen) [reply lib/subscriber.rb:43]: Test for queue url - for happy flow and errors

File: tests/publisher_spec.rb (hunk 1 above)
Comment (rocky02) [reply tests/publisher_spec.rb]: @glen its there in publisher_spec.rb:L9.. The title of the test is wrong. Correcting it now.

Comment (glen) [reply tests/publisher_spec.rb]: Meant that you also need to have a `be_falsey` test. So test with an invalid queue_name

**Comment by rocky02** [comment]
I am wondering how to test the send_messages / receive_messages aws sqs calls. Stubbing responses won't actually give a legit validation of the service. That's why was trying to configure elasticmq but facing issues. If this works out, aws.yml file and unnecessary loading of a few files will be handled.

**Comment by glen** [comment]
Go with stubs. Lets not over complicate matters by adding `elasticmq`.

File: specs/publisher_spec.rb [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    
+    it 'validates the sleep period' do
+      expect(Publisher.valid_sleep_period?(valid_sleep_period)).to be_truthy
+      expect { Publisher.valid_sleep_period?(lower_invalid_sleep_period) }.to raise_error(QItInvalidArgumentError)
+      expect { Publisher.valid_sleep_period?(upper_invalid_sleep_period) }.to raise_error(QItInvalidArgumentError)
+    end
+  end
+
+  context '#create_std_q' do
+  	let (:queue_name) { 'test123' }
```
Comment (glen) [inline specs/publisher_spec.rb]: Indentation

File: specs/publisher_spec.rb [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    let (:options) { [queue_name, 6] }
+    let (:publisher) { Publisher.new(options)}
+
+    it 'should read from queue' do
+    	result = publisher.create_std_q
```
Comment (glen) [inline specs/publisher_spec.rb]: Indentation

File: tests/publisher_spec.rb (hunk 1 above)
Comment (rocky02) [reply tests/publisher_spec.rb]: in case of invalid queue names, we raise error. Have added checks for those.

File: specs/publisher_spec.rb [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+      expect { Publisher.valid_queue_name?(invalid_queue_name2) }.to raise_error(QItInvalidArgumentError)
+    end
+  end
+
+  context '#valid_sleep_period?' do
+    let (:valid_sleep_period) { '7' }
+    let (:lower_invalid_sleep_period) { '2' }
+    let (:upper_invalid_sleep_period) { '21' }
+    
+    it 'validates the sleep period' do
```
Comment (glen) [inline specs/publisher_spec.rb]: Have separate context for `valid` and `invalid`.

File: specs/publisher_spec.rb [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    let (:invalid_queue_name1) { 'a'*81 }
+    let (:invalid_queue_name2) { 'testing/1@2' }
+    
+    it "should validate the queue name as per aws sqs format and return true for valid queue name" do
+      expect(Publisher.valid_queue_name?(valid_queue_name)).to be_truthy
+    end
+
+    it "should validate the queue name as per aws sqs format and raise error for invalid queue name" do
+      expect { Publisher.valid_queue_name?(invalid_queue_name1) }.to raise_error(QItInvalidArgumentError)
+      expect { Publisher.valid_queue_name?(invalid_queue_name2) }.to raise_error(QItInvalidArgumentError)
```
Comment (glen) [inline specs/publisher_spec.rb]: Good checking for invalid queue names

File: specs/publisher_spec.rb [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+  context '#valid_queue_name?' do
+    let (:valid_queue_name) { "test_123" }
+    let (:invalid_queue_name1) { 'a'*81 }
+    let (:invalid_queue_name2) { 'testing/1@2' }
+    
+    it "should validate the queue name as per aws sqs format and return true for valid queue name" do
+      expect(Publisher.valid_queue_name?(valid_queue_name)).to be_truthy
+    end
+
+    it "should validate the queue name as per aws sqs format and raise error for invalid queue name" do
```
Comment (glen) [inline specs/publisher_spec.rb]: Break this into 2 specs
```
[invalid_queue_name1, invalid_queue_name2].each do |invalid_queue_name|
  it "should validate the queue name as per aws sqs format and raise error for #{invalid_queue_name}" do
  ...
  ...
  end
end
```

Do a similar code for valid/invalid sleep period, queue urls

File: specs/spec_helper.rb [hunk 10]
```diff
@@ -0,0 +1,5 @@
+require 'rspec'
+require 'simplecov'
+SimpleCov.start
+require_relative '../lib/publisher'
+require_relative '../lib/subscriber'
```
Comment (glen) [inline specs/spec_helper.rb]: newline

**Comment by glen** [comment]
Ideally have a single expectation statement for each spec. Hence my pushing for having a loop around a spec with multiple values to test.
With multiple expectations - it gets difficult to understand which one fails when the spec fails.

**Comment by glen** [comment]
The specs for
```
rspec ./specs/subscriber_spec.rb:10
rspec ./specs/subscriber_spec.rb:42
```
are failing on my local system - this is because when defining the `valid_url` you are setting the name with a hard coded region. This will cause people using your repository and having accounts in a region other than `ap-south-1` to fail. Make it configurable - in fact you should ideally stub the region - so you don't even have to deal with the users input.

Also only ~ 78% of code is covered. Get it to 100%.

**Comment by rocky02** [comment]
Right now working on stubbing aws configurations and aws file creation part so it doesn't really have to hit sqs server. :+1:

**Comment by rocky02** [comment]
@glen we might need ENV now, we need to avoid calling `generate_aws_yml_file` and checking there for  the values in case of tests alone. Please suggest.
Also, I found [Stubbing for AWS Client](https://docs.aws.amazon.com/sdkforruby/api/Aws/ClientStubs.html) which is nice. Will try to implement it later.

**Comment by glen** [comment]
You can stub everything out.. yaml file and reading from it
` allow(File).to receive(:read).with(filename).and_return(file_json)`

The link you sent makes sense. Go ahead and use it.

**Comment by glen** [comment]
No need to stub out `ENV` - you can do without.

File: lib/aws_loader.rb [hunk 11]
```diff
@@ -25,4 +25,4 @@ def generate_aws_yml_file
   end
 rescue => e
   puts "Exception with aws.yml file #{e.inspect}"
-end
+end
```
Comment (glen) [inline lib/aws_loader.rb]: Newline

File: spec/publisher_spec.rb [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+    
+    [lower_invalid_sleep_period, upper_invalid_sleep_period].each do |invalid_sleep_period|
+      it "should validate the sleep period between 5-20 seconds and raise error for #{invalid_sleep_period}" do
+        expect { Publisher.valid_sleep_period?(invalid_sleep_period) }.to raise_error(QItInvalidArgumentError)
+      end
+    end
+  end
+
+  context '#create_std_q' do
+  	let (:queue_name) { 'test123' }
```
Comment (glen) [inline spec/publisher_spec.rb:36]: Indentation

**Comment by glen** [comment]
Shaping up well Neha. Few minor changes in this PR.
Some more specs for file read and create are pending. But overall good going.

**Comment by glen** [comment]
Mentioned this earlier - your test for validating the url is failing as it is reading from the yaml file. stub the reading and you should have a handle on it.

**Comment by glen** [comment]
When validating `queue_url` which depends on region - this can be different for each user. The best option is to stub the details. In you case the constant `AWS` that is being set. By doing something like `stub_const('AWS', {"access_key_id"=>"access_key", "secret_access_key"=>"secret_access_key", "region"=>"ap-south-1"})` this will ensure that you are not dependent on the users data values.
Adding @pcbishnoi

**Comment by rocky02** [comment]
@glen the suggestion you gave will not work, since its trying to load aws.yml file before that. Have already tried this solution using before(:suite) and it still does not work. Need to make some more changes to framework to prevent that piece of code being called.

File: lib/aws_loader.rb [hunk 13]
```diff
@@ -1,28 +1,33 @@
...
-AWS_PATH = File.join(QIt.root, 'config/aws.yml')
-
-def generate_aws_yml_file
-  sample_aws = {"aws"=>{"access_key_id"=>"", "secret_access_key"=>"", "region"=>""}}
-  File.open(AWS_PATH, 'w') do |file|
-    file.puts sample_aws.to_yaml
+module AwsLoader
+  
+  AWS_PATH = File.join(QIt.root, 'config/aws.yml')
+  AWS = YAML.load(File.read(AWS_PATH))["aws"] if File.exists?(AWS_PATH)
```
Comment (pcbishnoi) [inline lib/aws_loader.rb:6]: You're reading this twice (L18). Why so?

Comment (rocky02) [reply lib/aws_loader.rb:6]: module does not allow for dynamic constant assignment. Earlier when it was plain script, we were assigning the AWS constant there. But with the changes, need to assign the constant separately.

Comment (pcbishnoi) [reply lib/aws_loader.rb:6]: So, instead of reading the same thing again at L18, you can use this constant.

**Comment by pcbishnoi** [comment]
Neha, what's the code coverage now?

**Comment by rocky02** [comment]
Last I checked for the entire suite it was 89.5%. Adding module tests for aws_loader.rb currently.

**Comment by glen** [comment]
You may have to have a file similar as what I did for boot.rb that loads
the yaml files.
That will ensure that the boot process occurs through a single entry
point..

**Comment by rocky02** [comment]
@glen Did that already. Helped me solve the problem. The file name is application_config.rb. Might consider renaming it to boot.rb. The yml file load issue has been sorted though.

**Comment by glen** [comment]
Can you move the specs for the lib in it's own folder under spec - will be better organized.

**Comment by glen** [comment]
Added a new Issue for #18 logging to a file. Would make sure we do not get stdout when running tests.

**Comment by glen** [comment]
Have you written specs for[ aws_loader.rb](https://github.com/rocky02/q_it/blob/feature/addition_of_specs_and_tests/lib/aws_loader.rb) ? That is showing coverage of only 49% and dragging down your total coverage.
Screen Shot 2019-05-22 at 12 15 53 PM

If you have not done so - then we can do this spec activity in a new PR and close this one.

**Comment by rocky02** [comment]
@glen I have written the tests for aws_loader.rb. But I did not push that file. The tests there pass, but if I run the entire suite, it executes just that one test and stops and none of the other tests execute. Trying to fix that. Once that is resolved, I will push that up into this PR itself.

File: lib/aws_loader.rb [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+    rescue => e
+      puts "Exception with aws.yml file #{e.inspect}"
+    end
+  end
+end
+
+# Dummy class for RSpec testing purposes
+class ModuleTest
+  include AwsLoader
+end
```
Comment (glen) [inline lib/aws_loader.rb:38]: Do not add this here, but in the rspec test.

File: spec/aws_loader_spec.rb [hunk 15]
```diff
@@ -0,0 +1,25 @@
+RSpec.describe AwsLoader do
```
Comment (glen) [inline spec/aws_loader_spec.rb:1]: Add the class `ModuleTest` as a dummy class here above this line.
The ModuleTest includes the `aws_loader_spec` module

File: spec/aws_loader_spec.rb [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+
+  let (:mod_test) { ModuleTest.new }
+
+  context 'aws.yml file' do    
+    context '#configure_aws_file' do
+      it 'should return false for aws.yml file check and call #configure_aws_file and create a new .yml file' do
+        stub_const('AwsLoader::AWS_PATH', nil)
+        allow(File).to receive(:exists?).and_return(false)
+        expect(STDOUT).to receive(:puts).and_return('No `aws.yml` file present!')
+        expect(STDOUT).to receive(:puts).and_return('Created file and set the appropriate values!')
```
Comment (glen) [inline spec/aws_loader_spec.rb:11]: The above 2 expectations are syntactically incorrect. Should be 
`expect(STDOUT).to receive(:puts).with('No `aws.yml` file present!')` - we are expecting `puts` with a specific string.
Same as below test

**Comment by glen** [comment]
Merging this to develop. Additional changes will be completed in a new PR.

**Pull request merged**